src/checkers/inference/solver/backend/AbstractFormatTranslator.java

```diff
@@ -0,0 +1,161 @@
+        return null;
+    }
+
+    protected InequalityConstraintEncoder<ConstraintEncodingT> createInequalityConstraintEncoder(ConstraintVerifier verifier) {
+        return null;
+    }
+
+    protected ComparableConstraintEncoder<ConstraintEncodingT> createComparableConstraintEncoder(ConstraintVerifier verifier) {
+        return null;
+    }
+
+    protected CombineConstraintEncoder<ConstraintEncodingT> createCombineConstraintEncoder(ConstraintVerifier verifier) {
+        return null;
+    }
+
+    protected PreferenceConstraintEncoder<ConstraintEncodingT> createPreferenceConstraintEncoder(ConstraintVerifier verifier) {
+        return null;
+    }
+
+    protected ExistentialConstraintEncoder<ConstraintEncodingT> createExistentialConstraintEncoder(ConstraintVerifier verifier) {
+        return null;
+    }
+
+    /*Caution: if subclass overrides the serialize(Constraint) methods, nullness of the corresponding
+    * encoder field must be checked before dereferencing. Because subclass might not support some constraint
+    * encoding and the encoders of those unsupported constraints are by default null. */
+    @Override
+    public ConstraintEncodingT serialize(SubtypeConstraint constraint) {
+        return subtypeConstraintEncoder == null ? null :
+                ConstraintEncoderDispatcher.dispatch(constraint, subtypeConstraintEncoder);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(EqualityConstraint constraint) {
+        return equalityConstraintEncoder == null ? null :
+                ConstraintEncoderDispatcher.dispatch(constraint, equalityConstraintEncoder);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(InequalityConstraint constraint) {
+        return inequalityConstraintEncoder == null ? null :
+                ConstraintEncoderDispatcher.dispatch(constraint, inequalityConstraintEncoder);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(ComparableConstraint constraint) {
+        return comparableConstraintEncoder == null ? null :
+                ConstraintEncoderDispatcher.dispatch(constraint, comparableConstraintEncoder);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(PreferenceConstraint preferenceConstraint) {
+        return preferenceConstraint == null ? null : preferenceConstraintEncoder.encode(preferenceConstraint);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(CombineConstraint combineConstraint) {
+        return comparableConstraintEncoder == null ? null :
+                ConstraintEncoderDispatcher.dispatch(combineConstraint, combineConstraintEncoder);
+    }
+
+    @Override
+    public ConstraintEncodingT serialize(ExistentialConstraint constraint) {
+        return existentialConstraintEncoder == null ? null : existentialConstraintEncoder.encode(constraint);
+    }
+
+    @Override
+    public SlotEncodingT serialize(VariableSlot slot) {
```

Comment:
Mier, since `AbstractFormatTranslator` is an abstract class, maybe we can remove those empty `serialize(Slot)` methods? Or this is your intention to set the default behavior of serializing slots should return null?

Reply:
This is kind of default behaviour. I'm not quite clear about serializing slots. If we remove the default implementations, every subclass of FormatTranslator needs to implement those serialze(Slot) methods. Is this a normal scenario? If yes, we can remove them.

Reply:
I think each concrete FormatTranslator need to implement serialize(Slot) methods, as it implements `Serializer` interface, which requires both serialize(Constraint)` and `serialize(Slot)` methods should be implemented.

For MaxSat and LogiQL, I think they have special problems with implementing `serialize(Slot)` methods, so maybe at general level we require each concrete FormatTranslator have to implement all serialize methods, and if some sub-class doesn't support some methods, make it explicit return null in that sub-class.

Reply:
If we agree on this point, let me do this, as I currently working on this branch.

Reply:
NVM, I just realized we already have default `null` implementation on `serialize(Constriant)`, so maybe just keep the consistence in general level, return `null` on all `serialize` methods.

Reply:
Do you mean in MaxSAT and LogiQLFormatTranslator, all serialize(Slot) are explictily implemented to return null; In Z3FormatTranslator, serialize(Slot) methods are explicitly implemented with meaningful logic?

Reply:
Yes, that is my original thought. But I just realize since we already have `null` implementation in `serialize(Constraint)`, where the design for sub-class is "only implement methods that it supports". So current `serialize(Slot)` return null makes sense, as it keep consistent with the design of `serialize(Constraint)`, i.e. a sub-class "only implement serialize methods that it supports".

So then we just need make this design documented in the abstract class.

Reply:
Actually, in this part, I want to move encode(PreferenceConstraint) and encode(ExistentialConstraint) also into ConstraintEncoderDispatcher class but maybe using a new Name, e.g. ConstraintEncoderDirector

Reply:
Sure, I'll document this design decision in the documentation.

Reply:
Yes, that also improve the consistence. Agree on this point.

Reply:
This encapsulation is done.